### DISCUSSION TITLE
A research result on Performance Evaluation of OpenCIlk Program on Asymmetric Performance Multicore Processors

### DISCUSSION THREAD

**Discussion by adnan4791** [OpenCilk/opencilk-project#167] [Show and tell]
Dear Experts,

I have done a performance evaluation of the OpenCIlk Program on an Asymmetric Performance Multicore Processor, AMPs for short. AMPs, such as Intel Core i5 1240P, integrate high-performance Core (P-CPUs) and efficient Core (E-CPUs). Their performance is imbalanced. Specifically, the E-CPU is slower. Hence the first problem, in this case, is the unfair workload distribution among the P-CPUs and E-CPUs. Some literature supposed that AMPs might cause a performance bottleneck due to the completion time being bound to slower CPUs in static scheduling. In this case the faster CPUs would suffer from reduced performance (reduced IPC).
I propose to apply OpenCilk as a solution to the problem for several reasons. The first reason is that OpenCilk is suitable for shared memory multicores such as AMPs. Second, OpenCilk implements lazy task creation, which has low overhead. Third, OpenCilk implements continuation steal with THE protocol, so steal operations do not cause significant overhead. 
I evaluated the simple pi program, which has a regular workload, written in OpenCilk.
As expected, i found 
1. OpenCilk performs a simple pi program well.
2. P-CPUs does not suffer from reduced IPC in hybrid mode.
3. Only 1 E-CPU suffers from a little bit reduced IPC in hybrid mode.
4. Using Cilkscale tool, the performance can be sub linear or super linear, it depends on which CPU as the reference (T(1)).
5.  From the poin 4 , i would like to propose two updates to Amdahl's law equation.

 I would like to write a paper, in this regard, i would like ask your comments and advice, and i apologize for my bad English.

Edited : 
I am writing my questions here
1. Are there any related research findings from this community?
2. Are there any additional results that I need to add so that the paper will have enough contribution for the journal article?
3. If there is someone in this community who is conducting similar research? 

i would like to thanks in advance for this opportunity,  

adnan

**Comment by behoppe** [comment]
Hi @adnan4791, thanks for your post. I would like to clarify your question, so that one of the OpenCilk experts can better answer it. 

To reiterate your background information, it sounds like you have used OpenCilk to evaluate the performance of a test (simple pi) program with the goal of exploring the performance characteristics of AMPs, and based on that, you have arrived at two proposed updates to Amdahl's law.  

For your question, are there any kinds of comments or advice you're seeking? Based on other correspondence I've had with you, I think you might want specifically want to know if anyone in our community is doing similar work, or knows of other people or conferences that would be relevant. Is that right?

Best regards,
Bruce

**Reply by adnan4791** [reply]
Dear @behoppe and Experts 

Its me Adnan,

Thank you for your response. I would like to clarify my question. Before I ask, I would like to provide some background information on what I actually need.

The leadership of our university, including the rector and dean of the faculty, encourages us to write journal articles. We believe that publishing journal articles will improve our university's ranking. Therefore, we have an obligation to write and publish them.

As my background and interests are related to performance engineering on multicore processors, and I am intrigued by the concept of work-stealing, I am conducting research and writing a journal article on the topic. I am seeking information, as you mentioned, about whether there is any work related to work-stealing on AMPs in this community or if any members of this community are working on this issue. That is what I would like to ask.

Additionally, I have another question: Besides the results I have previously mentioned, are there any other ideas I should include to ensure that the work makes a significant contribution to a journal article?

thanks in advance

**Comment by behoppe** [comment]
Here is an applied question related to this thread, posted recently: [How well do Linux and Windows handle variation in CPU core performance?](https://news.ycombinator.com/item?id=36269244)

**Reply by adnan4791** [reply]
HI @behoppe 
Thank you for your information,
I think the operating system level load balancer works to balance the load of all the threads of multiple applications.
In applications with multiple threads, of course, the operating system does not know how the load situation of user-level threads is. In multi-threaded applications, generally using one to one mapping. So the loading situation of user-level threads reflects the loading situation of kernel threads. So that the user-level load balancer plays an important role in multi-threaded applications. In this case OpenCilk is also the best solution.

What are the opinions of the experts from this community? Actually I also hope for enlightenment from related experts regarding this.
